bgpy/simulation_engine/policies/aspa/aspa.py

```diff
@@ -40,104 +41,144 @@ def _valid_ann(self, ann: "Ann", from_rel: Relationships) -> bool:  # type: igno
         else:
             raise NotImplementedError("Should never reach here")
 
-    def _upstream_check(self, ann: "Ann", from_rel: Relationships) -> bool:
+    def _upstream_check(self, ann: "Ann", from_rel: "Relationships") -> bool:
         """ASPA upstream check"""
 
         # Upstream check
         if len(ann.as_path) == 1:
             return super()._valid_ann(ann, from_rel)
         else:
-            reversed_path = ann.as_path[::-1]  # type: ignore
             # For every adopting ASPA AS in the path,
             # The next ASN in the path must be in their providers list
-            for i in range(len(reversed_path) - 1):
-                if not self._provider_check(reversed_path[i], reversed_path[i + 1]):
-                    return False
 
-        return super()._valid_ann(ann, from_rel)  # type: ignore
+            # 4. If max_up_ramp < N, the procedure halts with the outcome "Invalid".
+            if self.get_max_up_ramp_length(ann) < len(ann.as_path):
+                return False
+
+            # No need to actually verify the min_up_ramp as we accept Unknown announcements
+            # 5. If min_up_ramp < N, the procedure halts with the outcome "Unknown".
+            if self.get_min_up_ramp_length(ann) < len(ann.as_path):
+                # Valid as we do not invalidate Unknown announcements
+                return super()._valid_ann(ann, from_rel)
```

Comment:
Do you think it may be possible to remove, or at least comment out, lines 60-62? Since we don't differentiate in this case between unknown and valid, we can just return valid I think? The ASPA function is just pretty slow in general so if we can remove an unnecessary function call that would be great :)